osu.Game/Extensions/DrawableExtensions.cs

```diff
@@ -56,12 +60,34 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
             component.Scale = info.Scale;
             component.Anchor = info.Anchor;
             component.Origin = info.Origin;
+            component.UsingClosestAnchor().Value = info.UsingClosestAnchor;
 
             if (component is Container container)
             {
                 foreach (var child in info.Children)
                     container.Add(child.CreateInstance());
             }
         }
+
+        /// <remarks>
+        /// <p>A <see cref="ConditionalWeakTable{TKey,TValue}">ConditionalWeakTable</see> is preferable to a <see cref="Dictionary{TKey,TValue}">Dictionary</see> because a <c>Dictionary</c> will keep
+        /// orphaned references to an <see cref="ISkinnableDrawable"/> forever, unless manually pruned.</p>
+        /// <p><see cref="BindableBool"/> is used as a thin wrapper around <see cref="System.Boolean">bool</see> because <c>ConditionalWeakTable</c> requires a reference type as both a key and a value.</p>
+        /// <p><see cref="IDrawable"/> was chosen over <see cref="Drawable"/> because it is a common ancestor between <see cref="Drawable"/> (which is required for <see cref="Drawable.Anchor"/> logic)
+        /// and <see cref="ISkinnableDrawable"/> (which is required for serialization via <see cref="SkinnableInfo"/>).</p>
+        /// <p>This collection is thread-safe according to the
+        /// <a href="https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-5.0#thread-safety">documentation</a>,
+        /// but the <c>BindableBool</c>s are not unless <see cref="Bindable{T}.BeginLease">leased</see>.</p>
+        /// </remarks>
+        private static readonly ConditionalWeakTable<IDrawable, BindableBool> is_drawable_using_closest_anchor_lookup = new ConditionalWeakTable<IDrawable, BindableBool>();
```

Comment:
> This cannot exist. Game-global dictionaries storing drawable references are a huge no-no. 

Makes sense.

> This needs to be localised more to the skin editor itself in some capacity.

Is this just a matter of finding where to put a `[Cached]` storage wrapper around the `ConditionalWeakTable`?

Reply:
Also, I'm not saying that `DrawableExtensions` is a good place for this—it's not, and I will change this to be osu-compliant. I understand and agree with you in principal that holding onto global references is bad. But I'm also having trouble seeing the pitfalls that you're suggesting are present in this specific implementation.

In its current form, `UsingClosestAnchor()` will never give back a reference to a `Drawable`; it only takes references in. If a key (a `Drawable`) in the table is not referenced externally, the entry is deleted by `ConditionalWeakTable` and garbage-collected. A `BindableBool` has no concept of a `Drawable`, so no references to it are held or exposed in that way.

So the worst-case scenario seems to be that two separate classes, which independently already have a reference to a given `Drawable`, are able to obtain a common `BindableBool` associated with it, but that `BindableBool` has no references to or from the original `Drawable`. So, as I pointed out in the xmldoc, there could be threading issues when writing to the underlying value of the shared `BindableBool`.

I'm only bringing this up so that I'm clear on the drawbacks in upcoming iterations.

Reply:
Caching the collection somewhere in the skin editor hierarchy would be the obvious way to proceed, yes.

As to the implementation as it reads currently: I did not mean to suggest (or even want to argue about) the correctness of the reference-holding logic. The reason I rejected it so sternly is quite a bit simpler than that:

1. This project has a hard-line stance against shared global state, and provides mutliple facilities and patterns designed to avoid keeping such state or helping manage it if it really does need to exist. I don't think I need to explain why shared global state is bad (very hard to debug, leads to "spooky action at a distance"). In this case it is not necessary to store global state, so it should not be done.
2. Having an extension method *store* global state is extremely non-idiomatic. I have been reading C# code for quite a few years now, and have never seen anything of the sort. What extension methods are usually intended for is usually small pieces of convenience code to extend a class that cannot be extended in another manner.

Reply:
That makes a lot of sense. Thanks for being patient with me.